apps/condo/domains/organization/utils/accessSchema.js

```diff
@@ -1,239 +1,308 @@
-
-    for (const id of organizationIds) {
-        const hasAccess = await checkPermissionInUserOrganizationOrRelatedOrganization(userId, id, permission)
-        if (!hasAccess) return false
+    for (const info of Object.values(userOrganizationsInfo.organizations)) {
+        if (_checkPermissionsInOrganization(info.permissions, permissionsToCheck)) {
+            organizationIds.push(...info.childOrganizations)
+        }
     }
-
-    return true
+    return uniq(organizationIds)
 }
 
-async function checkUserBelongsToOrganization (userId, organizationId) {
-    if (!userId || !organizationId) return false
-    const employee = await getByCondition('OrganizationEmployee', {
-        organization: { id: organizationId },
-        user: { id: userId },
-        isAccepted: true,
-        isBlocked: false,
-        isRejected: false,
-        deletedAt: null,
-    })
+/**
+ * Gets the IDs of organizations and its child (related) organizations in whose parent organizations the user is employed and has all permissions from the list.
+ * @param {{ id: string }} user - user object
+ * @param {Array<string> | string} permissions - permissions to check,
+ * can be passed as array of strings for multiple permissions or a single string for a single permission
+ * @returns {Promise<Array<string>>}
+ */
+async function getEmployedOrRelatedOrganizationsByPermissions (user, permissions) {
+    const userOrganizationsInfo = await _getUserOrganizations(user)
 
-    if (!employee) {
-        return false
-    }
+    const permissionsToCheck = Array.isArray(permissions) ? permissions : [permissions]
+    const organizationIds = []
 
-    if (employee.isBlocked) {
-        return false
+    for (const [id, info] of Object.entries(userOrganizationsInfo.organizations)) {
+        if (_checkPermissionsInOrganization(info.permissions, permissionsToCheck)) {
+            organizationIds.push(id, ...info.childOrganizations)
+        }
     }
+    return uniq(organizationIds)
+}
 
-    return employee.deletedAt === null
+/**
+ * Checks if user is employed in all listed organizations and has all correct permissions in it.
+ * Both organizations and permissions can be single elements if passed as strings instead of arrays
+ * This utils is faster than filtering organization ids from corresponding get<> function,
+ * so use it when you don't need related organizations
+ * @param {{ id: string }} user - user object
+ * @param {Array<string> | string} organizationIds - organizations to checks (can be passed as array of IDs or a single ID)
+ * @param {Array<string> | string} permissions - permissions to check (can be passed as array or a single string)
+ * @returns {Promise<boolean>}
+ */
+async function checkPermissionsInEmployedOrganizations (user, organizationIds, permissions) {
+    const userOrganizationsInfo = await _getUserOrganizations(user)
+
+    const organizationsToCheck = Array.isArray(organizationIds) ? uniq(organizationIds) : [organizationIds]
```

Comment:
^^ same for `user` and `organizations`

and for other utils

Reply:
I'm not sure, that I should do it. The reason behind empty organizations I've described below and what about user.

The goal of this util is to check if user have access to specified organizations, JSDoc added the interface that user is required parameter. So you must provide a user :)

(In some other typed language like Go (or even Typescript) you'll be punished for not passing a user, so if you create signature and passing something else - that's your fault)

Maybe I should add user-guard, but I'm not sure that I should do it. My point is that by moving more guards and logics to utils you're skipping some corner cases in access functions themself. 
(Because everyone will assume that your util is Swiss army knife :) )

Reply:
> (In some other typed language like Go (or even Typescript) you'll be punished for not passing a user, so if you create signature and passing something else - that's your fault)

JS does not have such checks, so developers should do this) 
Just in case, I would add a check so as not to worry about it)